File: Python/Easy/225.py
# Daily Coding Problem: 225
# Difficulty: Easy
# Asked by: Bloomberg
# Authored by: TenthEdict

# Time Complexity: O(n**2)
def josephus_On2(n, k):
    prisoners = list(range(1, n+1))
    index = k-1
    while len(prisoners) > 1:
        prisoners.pop(index)
        index = (index + k - 1) % len(prisoners)
    return prisoners[0]

print(josephus_On2(5,2)) # prints 3
print(josephus_On2(1,5)) # prints 1
print(josephus_On2(7,1)) # prints 7

# Time Complexity: O(log N)

def josephus_Ologn(n):
    p = 1
    while p * 2 <= n:
        p = p * 2
    index = 2 * (n - p) + 1
    return index 

print(josephus_Ologn(1)) # prints 1
print(josephus_Ologn(16)) # prints 1
print(josephus_Ologn(8)) # prints 1

for n in range(1, 17):
    slow = josephus_On2(n, 2)
    fast = josephus_Ologn(n)
    match = "✓" if slow == fast else "✗"
    print(f"n={n}: O(n²)={slow}, O(log n)={fast} {match}")

File: Rust/Easy/225.rs
// Daily Coding Problem: 225
// Difficulty: Easy
// Asked by: Bloomberg
// Authored by: TenthEdict

fn josephus_on2(n: usize, k: usize) -> usize {
    let mut prisoners: Vec<usize> = (1..=n).collect();
    let mut index: usize = k - 1;
    while prisoners.len() > 1 {
        prisoners.remove(index);
        index = (index + k - 1) % prisoners.len();
    }
    prisoners[0]
}

fn josephus_ologn(n: usize) -> usize {
    let mut p: usize = 1;
    while p * 2 <= n {
        p = p * 2;
    }
    2 * (n - p) + 1
}

fn main() {
    println!("Comparing O(n²) and O(log n) solutions for k=2:");
    for n in 1..=16 {
        let slow = josephus_on2(n, 2);
        let fast = josephus_ologn(n);
        let check = if slow == fast { "✓" } else { "✗" };
        println!("n={}: O(n²)={}, O(log n)={} {}", n, slow, fast, check);
    }
}